# Batch Normalization

## **Q1. Theory and Concepts**

## 1. Explain the concept of batch normalization in the context of Artificial Neural Networks.

-> Batch Normalization is used to improve the training process and performance of the artificial neural network by normalizing the input values to each layer.

-> It is also used to solve some problems like internal covariate shift.

* **There are two methods of applying batch normalization on data :**

1. First normalization on input data then compute the neuron.

2. First compute the neuron and then apply normalization on the output value.

## 2. Describe the benefits of using batch normalization during training.

1. **Reduces the Internal Covariate Shift**

-> It helps to stabalize the distribution by applying normalization on it.

2. **Gradient Flow**

-> It maintain the gradient flow, that helps to prevent issues like vanishing gradient.

3. **Accelerate Convergence**

-> It allows higher learning rate, so we can reduce the number of epochs and can accelerate the convergence.

## 3. Discuss the working principle of batch normalization, including the normalization step and the learnable parameters.

* **Normalization Steps**

**Step-1 : Compute batch statistics**

mean = 1/n * sum(Xi)

stddev = np.sqrt(sum(Xi - mean) / (n - 1))

**Step-2 : Normalize Activations**

Standard Normalization(X) : (Xi - mean) / stddev

* **Learnable Parameters**

-> Apply linear transformation on normalized activation by the learnable parameters 𝛾(scale) and 𝛽(shift) :
 
   Yij = 𝛾Xi + 𝛽, where  
   
   𝛾 : Controls the scaling of the normalized activation
   
   𝛽 : Controls the shifting of the normalized activation

## **Q2. Implementation**

In [1]:
!pip install tensorflow

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 601.3/601.3 MB 2.7 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.7/133.7 kB 19.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.1/5.1 MB 60.7 MB/s eta 0:00:0000:01:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 53.7 MB/s eta 0:00:0000:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.3/5.3 MB 67.3 MB/s eta 0:00:00:00:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 54.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.5/65.5 kB 10.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 69.1 MB/s eta 0:00:00:00:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 80.3/80.3 kB 12.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.5/24.5 MB 43.3 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.5/57.5 kB 10.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 68.0

In [2]:
import warnings
warnings.filterwarnings('ignore')

import tensorflow as tf

In [3]:
mnist = tf.keras.datasets.mnist

In [4]:
(x_train,y_train),(x_test,y_test) = mnist.load_data()

In [5]:
x_test = x_test / 255
x_valid,x_train = x_train[:5000]/255 , x_train[5000:]/255
y_valid,y_train = y_train[:5000] , y_train[5000:]

## Without Normalization

In [6]:
LAYERS = [
        tf.keras.layers.Flatten(input_shape=(28,28)),
        tf.keras.layers.Dense(300,activation='relu'),
        tf.keras.layers.Dense(100,activation='relu'),
        tf.keras.layers.Dense(10,activation='softmax'),
        ]

In [7]:
model = tf.keras.models.Sequential(LAYERS)

In [8]:
model.compile(loss='sparse_categorical_crossentropy',optimizer='adam',metrics=['accuracy'])

In [9]:
model = model.fit(x_train,y_train,validation_data=(x_valid,y_valid),epochs=7)

Epoch 1/7
1719/1719 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.8862 - loss: 0.3748 - val_accuracy: 0.9706 - val_loss: 0.0971
Epoch 2/7
1719/1719 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9736 - loss: 0.0870 - val_accuracy: 0.9716 - val_loss: 0.0938
Epoch 3/7
1719/1719 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9822 - loss: 0.0563 - val_accuracy: 0.9802 - val_loss: 0.0682
Epoch 4/7
1719/1719 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9860 - loss: 0.0426 - val_accuracy: 0.9772 - val_loss: 0.0783
Epoch 5/7
1719/1719 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9897 - loss: 0.0320 - val_accuracy: 0.9768 - val_loss: 0.0809
Epoch 6/7
1719/1719 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9916 - loss: 0.0249 - val_accuracy: 0.9778 - val_loss: 0.0754
Epoch 7/7
1719/1719 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9938 - loss: 0.0194 - val_accuracy: 0.9820 - val_loss: 0.0750


## With Normalization

In [10]:
LAYERS = [
        tf.keras.layers.Flatten(input_shape=(28,28)),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(300,activation='relu'),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(100,activation='relu'),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(10,activation='softmax')
        ]

In [11]:
model_with_bn = tf.keras.models.Sequential(LAYERS)

In [12]:
model_with_bn.compile(loss='sparse_categorical_crossentropy',optimizer='adam',metrics=['accuracy'])

In [13]:
MODEL_WITH_BN = model_with_bn.fit(x_train,y_train,validation_data=(x_valid,y_valid),epochs=7)

Epoch 1/7
1719/1719 ━━━━━━━━━━━━━━━━━━━━ 12s 5ms/step - accuracy: 0.8931 - loss: 0.3446 - val_accuracy: 0.9656 - val_loss: 0.1096
Epoch 2/7
1719/1719 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9654 - loss: 0.1078 - val_accuracy: 0.9710 - val_loss: 0.1027
Epoch 3/7
1719/1719 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9724 - loss: 0.0827 - val_accuracy: 0.9756 - val_loss: 0.0890
Epoch 4/7
1719/1719 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9796 - loss: 0.0626 - val_accuracy: 0.9764 - val_loss: 0.0872
Epoch 5/7
1719/1719 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9834 - loss: 0.0510 - val_accuracy: 0.9784 - val_loss: 0.0953
Epoch 6/7
1719/1719 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9863 - loss: 0.0427 - val_accuracy: 0.9784 - val_loss: 0.0963
Epoch 7/7
1719/1719 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9885 - loss: 0.0373 - val_accuracy: 0.9774 - val_loss: 0.0995


**Model without Batch Normalization : Accuracy = 0.9938 & Loss = 0.0194** 

**Model with Batch Normalization : Accuracy = 0.9885 & Loss = 0.0373**

## **Q3. Experiments and Analysis**

## 1. Experiment with different batch sizes and observe the effect on the training dynamics and model performance.

* Batch Size = 33 | Accuracy = 0.9873

In [14]:
del model_with_bn

In [15]:
LAYERS = [
        tf.keras.layers.Flatten(input_shape=(28,28)),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(300,activation='relu'),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(100,activation='relu'),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(10,activation='softmax')
        ]

model_33 = tf.keras.models.Sequential(LAYERS)

model_33.compile(loss='sparse_categorical_crossentropy',optimizer='adam',metrics=['accuracy'])

MODEL = model_33.fit(x_train,y_train,validation_data=(x_valid,y_valid),epochs=7,batch_size=33)

Epoch 1/7
1667/1667 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.8890 - loss: 0.3505 - val_accuracy: 0.9662 - val_loss: 0.1118
Epoch 2/7
1667/1667 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9620 - loss: 0.1206 - val_accuracy: 0.9726 - val_loss: 0.0898
Epoch 3/7
1667/1667 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9758 - loss: 0.0760 - val_accuracy: 0.9720 - val_loss: 0.1042
Epoch 4/7
1667/1667 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9796 - loss: 0.0640 - val_accuracy: 0.9776 - val_loss: 0.0838
Epoch 5/7
1667/1667 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9830 - loss: 0.0511 - val_accuracy: 0.9754 - val_loss: 0.1000
Epoch 6/7
1667/1667 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9853 - loss: 0.0439 - val_accuracy: 0.9732 - val_loss: 0.0940
Epoch 7/7
1667/1667 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9873 - loss: 0.0391 - val_accuracy: 0.9800 - val_loss: 0.0807


In [16]:
del model_33

* Batch Size = 333 | Accuracy = 0.9976

In [18]:
LAYERS = [
        tf.keras.layers.Flatten(input_shape=(28,28)),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(300,activation='relu'),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(100,activation='relu'),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(10,activation='softmax')
        ]

model_333 = tf.keras.models.Sequential(LAYERS)

model_333.compile(loss='sparse_categorical_crossentropy',optimizer='adam',metrics=['accuracy'])

MODEL = model_333.fit(x_train,y_train,validation_data=(x_valid,y_valid),epochs=7,batch_size=333)

Epoch 1/7
166/166 ━━━━━━━━━━━━━━━━━━━━ 4s 14ms/step - accuracy: 0.8346 - loss: 0.5533 - val_accuracy: 0.9488 - val_loss: 0.2543
Epoch 2/7
166/166 ━━━━━━━━━━━━━━━━━━━━ 2s 12ms/step - accuracy: 0.9714 - loss: 0.1024 - val_accuracy: 0.9716 - val_loss: 0.1047
Epoch 3/7
166/166 ━━━━━━━━━━━━━━━━━━━━ 2s 12ms/step - accuracy: 0.9844 - loss: 0.0579 - val_accuracy: 0.9764 - val_loss: 0.0804
Epoch 4/7
166/166 ━━━━━━━━━━━━━━━━━━━━ 2s 12ms/step - accuracy: 0.9914 - loss: 0.0346 - val_accuracy: 0.9788 - val_loss: 0.0787
Epoch 5/7
166/166 ━━━━━━━━━━━━━━━━━━━━ 2s 12ms/step - accuracy: 0.9945 - loss: 0.0232 - val_accuracy: 0.9788 - val_loss: 0.0798
Epoch 6/7
166/166 ━━━━━━━━━━━━━━━━━━━━ 2s 12ms/step - accuracy: 0.9967 - loss: 0.0144 - val_accuracy: 0.9768 - val_loss: 0.0806
Epoch 7/7
166/166 ━━━━━━━━━━━━━━━━━━━━ 2s 12ms/step - accuracy: 0.9976 - loss: 0.0113 - val_accuracy: 0.9778 - val_loss: 0.0831


In [19]:
del model_333

* Batch Size = 3333 | Accuracy = 0.9792

In [20]:
LAYERS = [
        tf.keras.layers.Flatten(input_shape=(28,28)),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(300,activation='relu'),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(100,activation='relu'),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(10,activation='softmax')
        ]

model_3333 = tf.keras.models.Sequential(LAYERS)

model_3333.compile(loss='sparse_categorical_crossentropy',optimizer='adam',metrics=['accuracy'])

MODEL = model_3333.fit(x_train,y_train,validation_data=(x_valid,y_valid),epochs=7,batch_size=3333)

Epoch 1/7
17/17 ━━━━━━━━━━━━━━━━━━━━ 3s 54ms/step - accuracy: 0.5724 - loss: 1.3792 - val_accuracy: 0.7076 - val_loss: 1.3246
Epoch 2/7
17/17 ━━━━━━━━━━━━━━━━━━━━ 1s 38ms/step - accuracy: 0.9158 - loss: 0.2951 - val_accuracy: 0.8514 - val_loss: 0.9724
Epoch 3/7
17/17 ━━━━━━━━━━━━━━━━━━━━ 1s 37ms/step - accuracy: 0.9406 - loss: 0.2058 - val_accuracy: 0.8988 - val_loss: 0.7819
Epoch 4/7
17/17 ━━━━━━━━━━━━━━━━━━━━ 1s 38ms/step - accuracy: 0.9579 - loss: 0.1534 - val_accuracy: 0.9072 - val_loss: 0.6635
Epoch 5/7
17/17 ━━━━━━━━━━━━━━━━━━━━ 1s 38ms/step - accuracy: 0.9648 - loss: 0.1281 - val_accuracy: 0.9138 - val_loss: 0.5678
Epoch 6/7
17/17 ━━━━━━━━━━━━━━━━━━━━ 1s 36ms/step - accuracy: 0.9727 - loss: 0.1045 - val_accuracy: 0.9076 - val_loss: 0.5023
Epoch 7/7
17/17 ━━━━━━━━━━━━━━━━━━━━ 1s 35ms/step - accuracy: 0.9792 - loss: 0.0840 - val_accuracy: 0.8850 - val_loss: 0.4648


In [21]:
del model_3333

## 2. Discuss the advantages and potential limitations of batch normalization in improving the training of neural networks.

* Advantages

1. **Internel Covariate Shift Reduction** : By normalizing the distributed data batch normalization reduces the internal covariate shift which makes the training process more stable.

2. **Faster Convergence** : Batch normalization allows higher learning rate which leads the faster convergence.

3. **Overfitting** : It also helps to reduce the overfitting.

* Disadvantages

1. **Computational Overhead** : Batch Normalization add additional computational steps which increases the computational time and consumes more computational resources.

2. **Small Batch Size** : It can lead more noisy estimation of mean and variance which can affect the effectiveness of the normalization.

3. **** :